### PR TITLE
integrate keccak2

### PR DESCRIPTION
**Pull request by SchmErik**
still a WIP: marking as draft now and will mark as ready when tests/build errors are fixed

### CODE REVIEW

File: risc0/zkvm/src/host/protos/api.proto [hunk 1]
```diff
@@ -60,11 +61,10 @@ message ProveSegmentReply {
...
 
 message ProveZkrRequest {
-  base.Digest claim_digest = 1;
-  base.Digest control_id = 2;
-  bytes input = 3;
+  base.Digest control_id = 1;
+  bytes input = 2;
   // This is optional in the context of a CoprocessorRequest
-  AssetRequest receipt_out = 4;
+  AssetRequest receipt_out = 3;
```
Comment (flaub) [inline risc0/zkvm/src/host/protos/api.proto:67]: The ordinals are not allowed to change to maintain compatibility.

File: risc0/zkvm/src/host/server/session.rs [hunk 2]
```diff
@@ -93,6 +93,10 @@ pub struct Session {
     /// A list of pending ZKR proof requests.
     // TODO: make this scalable so we don't OOM
     pub(crate) pending_zkrs: Vec<ProveZkrRequest>,
+
+    /// A list of pending ZKR proof requests.
```
Comment (flaub) [inline risc0/zkvm/src/host/server/session.rs:97]:
```suggestion
    /// A list of pending Keccak proof requests.
```

File: risc0/circuit/keccak/Cargo.toml [hunk 3]
```diff
@@ -34,6 +34,7 @@ risc0-circuit-keccak-methods = { path = "methods" }
 risc0-zkvm = { workspace = true, features = [
   "std",
   "unstable",
+  "prove",
```
Comment (flaub) [inline risc0/circuit/keccak/Cargo.toml:37]: If we always assume `prove`, then do we still need a feature flag for this crate?

File: risc0/zkvm/src/host/client/env.rs [hunk 4]
```diff
@@ -79,6 +79,9 @@ pub struct ProveZkrRequest {
 /// A Keccak proof request.
 #[stability::unstable]
 pub struct ProveKeccakRequest {
+    /// The digest of the claim that this ZKR program is expected to produce.
+    pub claim_digest: Digest,
```
Comment (flaub) [inline risc0/zkvm/src/host/client/env.rs:83]: Why is this needed for a proof request?

File: risc0/circuit/keccak/Cargo.toml (hunk 3 above)
Comment (SchmErik) [reply risc0/circuit/keccak/Cargo.toml:37]: yes - the prove flag is set for dev-dependencies - I changed the test for this crate to not use coprocessor so it exercises the pending_keccaks code path (which is similar to the pending_zkrs thing we do) in the server.

File: risc0/zkvm/src/host/client/env.rs (hunk 4 above)
Comment (SchmErik) [reply risc0/zkvm/src/host/client/env.rs:83]: originally, I didn't think this was needed either. so I left it out. As i worked on the prover example, I saw that for prove_zkr in the prover example, this was utilized. I was thinking that this could be computed from the receipt produced from the proof but i saw that the prove_zkr did it thsi way so I added this. This is set by the host syscall handler and is passed from the guest. If there's a mismatch then it means that the guest and the prover disagreed on the digest, which is a good check to have. but I guess this spurs another question: why does this exist on the prove_zkr side?
